app/src/org/commcare/dalvik/activities/CommCareWiFiDirectActivity.java

```diff
@@ -175,9 +139,50 @@ public void onClick(View v) {
         mIntentFilter.addAction(WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION);
         mIntentFilter.addAction(WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION);
         mIntentFilter.addAction(WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION);
+        setupGridView();
         changeState();
+        if(savedInstanceState == null){
+            showChangeStateDialog();
+        }
     }
 
+    public void showChangeStateDialog(){
+        showDialog(this, localize("wifi.direct.change.state.title").toString(),
+                localize("wifi.direct.change.state.text").toString());
+    }
+
+    private void setupGridView() {
+        final RecyclerView grid = (RecyclerView)findViewById(R.id.wifi_direct_gridview_buttons);
+        grid.setHasFixedSize(true);
+
+        StaggeredGridLayoutManager gridView =
+                new StaggeredGridLayoutManager(3, StaggeredGridLayoutManager.VERTICAL);
+        grid.setLayoutManager(gridView);
+        grid.setItemAnimator(null);
+        grid.setAdapter(adapter);
+
+        grid.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {

```

Comment:
Not positive, but I don't think this listener is needed. @ctsims any reason this should be here? And can you give context as to why it is needed on the home screen adapter? Can it be removed there too?